### PR TITLE
rpc: implement whitelist for commands needed by Dash Platform

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
This is the first step of the features described here: https://docs.google.com/document/d/15Q4Y4sEk-t4Qm8cUnLi4btBg5HGHPIRIerY70g9G3jY/edit#

This change adds logic which enables the operator of a MN to restrict the rpcuser used by Dash Platform to the very specific subset of commands absolutely needed by Dash Platform. This is important for a number of reasons pointed out in the document above to ensure that a bug or vulnerability in Dash Platform does not have significant effects on the L1 Dash network.

This additional RPC user should be set as shown in the attached test.

### CODE REVIEW

File: src/rpc/server.cpp [hunk 1]
```diff
@@ -546,6 +556,25 @@ UniValue CRPCTable::execute(const JSONRPCRequest &request) const
...
+            // Check if this request matches a valid masternode rpc command
+            if (request.strMethod == cmd.first) {
+                // First arg / command matched, if there is no subcommand return true
+                if (cmd.second.empty()) return true;
+                if (request.params[0].isStr() && request.params[0].getValStr() == cmd.second) {
+                    // Sub command was needed and matched, return true
+                    return true;
+                }
+            }
+            return false;
```
Comment (UdjinM6) [inline src/rpc/server.cpp:571]: Too many if-s imo :) maybe smth like this
```suggestion
            // Check if this request matches a valid platform allowed rpc command
            return (request.strMethod == cmd.first) &&
                // and there is no subcommand
                (cmd.second.empty() ||
                // or a subcommand was needed and it matches too
                (request.params[0].isStr() && request.params[0].getValStr() == cmd.second));
```
instead?

File: src/rpc/server.cpp [hunk 2]
```diff
@@ -34,6 +34,16 @@ static RPCTimerInterface* timerInterface = nullptr;
 /* Map of name to timer. */
 static std::map<std::string, std::unique_ptr<RPCTimerBase> > deadlineTimers;
 
+// Any commands submitted by this user will have their commands filtered based on the platformAllowedCommands
+static std::string defaultPlatformUser = "platform-user";
```
Comment (UdjinM6) [inline src/rpc/server.cpp:38]:
```suggestion
static const std::string defaultPlatformUser = "platform-user";
```

File: src/rpc/server.cpp (hunk 1 above)
Comment (PastaPastaPasta) [reply src/rpc/server.cpp:571]: Hmm, idk, I find the logic flow of the if statements easier to understand

Comment (PastaPastaPasta) [reply src/rpc/server.cpp:571]: Nah, nevermind, I think this change makes sense :+1:

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by xdustinface — CHANGES_REQUESTED** [review]
See https://github.com/xdustinface/dash/tree/suggestions-3738 for some thoughts i had while reviewing.

**Comment by xdustinface** [comment]
@PastaPastaPasta @UdjinM6 i added three more commits there.

@UdjinM6 About your comment with the complex `if` thoughts about 5b7958d11a9df7e5fdc11d722528cba8a1603cce ?

e419f18c31e3333438c3da9c8f0e8451b7385e8e: I think it makes sense to rename it but if we want to be more general and stick with `RPC_PROTECTED_COMMAND` we should probably adjust the related comment in `protocol.h`?

**Comment by PastaPastaPasta** [comment]
@xdustinface test rpc_platform_filter.py doesn't pass, and I don't think I messed it up

**Comment by xdustinface** [comment]
> @xdustinface test rpc_platform_filter.py doesn't pass

Where? 🤔

**Comment by shumkov** [comment]
What if we just add an ability to define any users with a whitelist (and other limitations in the future) instead of introducing the specific one for the platform. I think it's a more flexible approach and foundation for future improvements. It will be similar to database user (i.e. MySQL) concept.

In the platform, we have different areas of code where we would like to use different limitations for RPC. For example, Tx Filter Stream can produce a lot of calls and it could lead to overloading of RPC queue so other important parts of the platform logic will just be stuck because of that. So in the future, besides command whitelist, it would be nice to have different RPC queues or some else that could prevent the single RPC queue from overloading by a specific user. 

Also, I believe in the future we will see other L2 solutions on top of Core blockchain so it will be useful for them too. But having a hardcoded user specifically for the platform kinda breaks layers design paradings where lower lever doesn't know about higher levels and could lead to some circular dependencies.

**Comment by PastaPastaPasta** [comment]
Note @shumkov, you can set a specific username as the platform user by using "-platform-user" launch param

Test is passing locally, probably forgot to compile :shrug: 

@UdjinM6 @xdustinface please review

File: src/rpc/server.cpp [hunk 3]
```diff
@@ -546,6 +556,22 @@ UniValue CRPCTable::execute(const JSONRPCRequest &request) const
     if (!pcmd)
         throw JSONRPCError(RPC_METHOD_NOT_FOUND, "Method not found");
 
+    // Before executing the RPC Command, filter commands from platform rpc user
+    if (fMasternodeMode && request.authUser == gArgs.GetArg("-platform-user", defaultPlatformUser)) {
```
Comment (UdjinM6) [inline src/rpc/server.cpp:560]: This probably requires a help string in `init.cpp`

**Review by UdjinM6 — APPROVED** [review]
re-utACK

**Review by xdustinface — COMMENTED** [review]
Hmmm i agree with @shumkov..  it would make more sense to implement it like a dynamic whitelist where you can provide the allowed parameter in the config. This way it would not require us to push changes and a new version if platform needs changes to the allowed RPC set.. they could just adjust their config and we could give all this a more generic naming. Shouldn't we just add this here instead of merging this now?

**Review by xdustinface — APPROVED** [review]
utACK because it seems like no one else cares about doing it more generic 🙈

**Pull request merged**